src/probnum/randprocs/_gaussian_process.py

```diff
@@ -0,0 +1,110 @@
+"""Gaussian processes."""
+
+from typing import Callable, Optional, Union
+
+import numpy as np
+
+from probnum import kernels, randvars
+from probnum.type import IntArgType, RandomStateArgType, ShapeArgType
+
+from ._random_process import RandomProcess
+
+_InputType = Union[np.floating, np.ndarray]
+_OutputType = Union[np.floating, np.ndarray]
+
+
+class GaussianProcess(RandomProcess[_InputType, _OutputType]):
+    """Gaussian processes.
+
+    A Gaussian process is a continuous stochastic process which if evaluated at a
+    finite set of inputs returns a random variable with a normal distribution. Gaussian
+    processes are fully characterized by their mean and covariance function.
+
+    Parameters
+    ----------
+    input_dim :
+        Shape of the input of the Gaussian process.
+    output_dim :
+        Shape of the output of the Gaussian process.
+    mean :
+        Mean function.
+    cov :
+        Covariance function or kernel.
+
+    See Also
+    --------
+    RandomProcess : Random processes.
+    GaussMarkovProcess : Gaussian processes with the Markov property.
+
+    Examples
+    --------
+    Define a Gaussian process with a zero mean function and RBF kernel.
+
+    >>> import numpy as np
+    >>> from probnum.kernels import ExpQuad
+    >>> from probnum.randprocs import GaussianProcess
+    >>> mu = lambda x : np.zeros_like(x)  # zero-mean function
+    >>> k = ExpQuad(input_dim=1)  # RBF kernel
+    >>> gp = GaussianProcess(mu, k)
+
+    Sample from the Gaussian process.
+
+    >>> x = np.linspace(-1, 1, 5)[:, None]
+    >>> np.random.seed(42)
+    >>> gp.sample(x)
+    array([[-0.35187364],
+           [-0.41301096],
+           [-0.65094306],
+           [-0.56817194],
+           [ 0.01173088]])
+    >>> gp.cov(x)
+    array([[1.        , 0.8824969 , 0.60653066, 0.32465247, 0.13533528],
+           [0.8824969 , 1.        , 0.8824969 , 0.60653066, 0.32465247],
+           [0.60653066, 0.8824969 , 1.        , 0.8824969 , 0.60653066],
+           [0.32465247, 0.60653066, 0.8824969 , 1.        , 0.8824969 ],
+           [0.13533528, 0.32465247, 0.60653066, 0.8824969 , 1.        ]])
+    """
+
+    def __init__(
+        self,
+        mean: Callable[[_InputType], _OutputType],
+        cov: Union[Callable[[_InputType], _OutputType], kernels.Kernel],
```

Comment:
how much do we want to enforce that a user provides a `Kernel` object as a covariance? If for instance the type would only be `kernels.Kernel` then callable code would still work (with some sort of duck-typing) but not force us into any promises (or, alternatively, we could wrap a callable into a kernel internally).

Reply:
If we only require a kernel, it would also remove the need for providing `input_dim` and `output_dim` and checks that they align with the kernel. 
I am not saying this is the way to go, but maybe something to think about (if you have thought about it already, then this will be easy to resolve :) )?

Reply:
I agree with forcing the user to implement their own `Kernel`. That way we can rely on the interface. Anything else seems unnecessarily complicated.